src/opentracing-c/tracer.c

```diff
@@ -0,0 +1,219 @@
+        (opentracing_destructible*) &((null_span*) span)->null_span_context;
+    span_context->destroy(span_context);
+
+    free(span);
+}
+
+static void null_span_finish(opentracing_span* span)
+{
+    (void) span;
+}
+
+static void
+null_span_finish_with_options(opentracing_span* span,
+                              const opentracing_finish_span_options* options)
+{
+    (void) span;
+    (void) options;
+    (void) span;
+}
+
+static opentracing_span_context* null_span_span_context(opentracing_span* span)
+{
+    assert(span != NULL);
+    return &((null_span*) span)->null_span_context;
+}
+
+static opentracing_bool null_span_set_operation_name(opentracing_span* span,
+                                                     const char* operation_name)
+{
+    (void) span;
+    (void) operation_name;
+    return opentracing_true;
+}
+
+static opentracing_bool null_span_set_tag(opentracing_span* span,
+                                          const char* key,
+                                          const opentracing_value* value)
+{
+    (void) span;
+    (void) key;
+    (void) value;
+    return opentracing_true;
+}
+
+static opentracing_bool null_span_log_fields(
+    opentracing_span* span, const opentracing_log_field* fields, int num_fields)
+{
+    (void) span;
+    (void) fields;
+    (void) num_fields;
+    return opentracing_true;
+}
+
+static opentracing_bool null_span_set_baggage_item(opentracing_span* span,
+                                                   const char* key,
+                                                   const char* value)
+{
+    (void) span;
+    (void) key;
+    (void) value;
+    return opentracing_true;
+}
+
+static const char* null_span_baggage_item(const opentracing_span* span,
+                                          const char* key)
+{
+    (void) span;
+    (void) key;
+    static const char* empty_str = "";
+    return empty_str;
+}
+
+static opentracing_tracer* null_span_tracer(opentracing_span* span)
+{
+    assert(span != NULL);
+    return ((null_span*) span)->null_tracer;
+}
+
+#define NULL_SPAN_INIT                   \
+    {                                    \
+        {{&null_span_destroy},           \
+         &null_span_finish,              \
+         &null_span_finish_with_options, \
+         &null_span_span_context,        \
+         &null_span_set_operation_name,  \
+         &null_span_set_tag,             \
+         &null_span_log_fields,          \
+         &null_span_set_baggage_item,    \
+         &null_span_baggage_item,        \
+         &null_span_tracer},             \
+            NULL_SPAN_CONTEXT_INIT, NULL \
+    }
+
+static opentracing_span* null_tracer_start_span_with_options(
+    opentracing_tracer* tracer,
+    const char* operation_name,
+    const opentracing_start_span_options* options)
+{
+    (void) tracer;
+    (void) operation_name;
+    (void) options;
+    null_span* span = (null_span*) malloc(sizeof(null_span));
```

Comment:
Does the noop tracer even need to allocate memory? Couldn't it just return a pointer to a single instance of null_span?

Reply:
In this case yes, and I can change it, as long as that wouldn't be the case for a real tracer.

I was tempted to make advertise this API as not allocating any heap memory, but that wouldn't be true for any instrumented process using a tracer. The only way around that is essentially capping the size of each object and reserving space for it in a stack buffer. See `sockaddr_storage` details [here](http://pubs.opengroup.org/onlinepubs/009696699/basedefs/sys/socket.h.html) for a similar idea with IPv4 and IPv6 sockets.